### ISSUE TITLE
Miniscript support?

### ISSUE THREAD

**Issue by joemphilips** [MetacoSA/NBitcoin#668]
Hi

I've been trying to port [rust-miniscript](https://github.com/apoelstra/rust-miniscript) into F#

https://github.com/joemphilips/FNBItcoin/tree/master/src/FNBitcoin 

My original idea is to make this library as a wrapper for NBitcoin, but I'm wondering if this should be included in this repository when it is finished. Any opinion ?

(It is almost done besides 1. decompiler is not finished, and 2. want to add more property tests to asssure the behaviour will be exactly the same with rust-miniscript in all case. )

reasons to merge into this repo is
1. more fluent `ScriptBuilder` class will be possible
2. PSBT could support arbitrary script for finalization (currently it is only for simple multisig)
3. PSBT can parse the script into human readable format and users could see if it is in the right form

reasons to `not` merge is
1. It is written in F# and does not align well with other source code.
  * I have no clue how to write this kind of software (parser, compiler) with C#
2. spec of miniscript itself is not formalized and might change slightly in the future
  * I don't think this is a big problem since 1) optimizing the compilation process seems to have done well and more optimization sounds like over-engineering. 2) it is likely to change in the future anyway, when MuSig and Graftroot comes out.

**Comment by joemphilips** [comment]
Oh yeah, one more reason to not include is

1. If it is possible to compile rust-miniscript to wasm and load that directly, that will be much better in any sense.

**Comment by NicolasDorier** [comment]
I think making a library for miniscript in F# or C# is a great idea. The logic should not be that hard, and mistakes are not so hard to find and damaging as if it was a cryptographic mistake so the WASM route seems not necessary.

That say, I would advise to add it in the main NBitcoin project instead of separate library or dependency.
Converting F# to C# can be done 95% via automated tools. (one solution, which is probably not the best, is use ilspy to decompile the F# assembly and convert the IL to C#)

I am quite sure I can find around converters that work at the abstract syntax tree level instead of IL level.

If you want I can do it once you think your library is complete.

**Comment by NicolasDorier** [comment]
Another idea actually which would work great, is that you just use extensions method in your MiniScript library.

By doing this you can "add helper methods" to Script / PSBT without changing NBitcoin.

**Comment by knocte** [comment]
> Converting F# to C# can be done 95% via automated tools

Why are you mentioning this? what's the point of converting F# to C#?

**Comment by NicolasDorier** [comment]
@knocte for including Miniscript directly into NBitcoin project, without any dependency.

**Comment by knocte** [comment]
Mmm, sorry but I still don't find the point. I assumed the N in "NBitcoin" stands for ".NET", not for C#. An F# library in a .NET nuget package wouldn't hinder it or add dependencies.

**Comment by NicolasDorier** [comment]
I mean I am pretty happy with NBitcoin being 1 big library as opposed to 35 decoupled packages. 
It makes so much things simpler.

So I don't want NBitcoin library to have a direct dependency to a MiniScript nuget library.
There is lot's of cool methods that can added to NBitcoin with MiniScript, but because NBitcoin will not reference the MiniScript package (and can't because this would be a circular dependency), then the best way seems to include MiniScript directly in C# inside NBitcoin.

Extensions method at Miniscript level are good, but the problem is that people referencing NBitcoin have no way to know it exists.

**Comment by NicolasDorier** [comment]
There is no good reason to include the Miniscript F# library directly into NBitcoin IMHO.

This will not increase discovery of Miniscript library, while needlessly blocking @joemphilips progress as I will need to manage releases.

**Comment by knocte** [comment]
Nuget packages can supply more than 1 library (dll). Then it's just a matter of namespace discoverability, just as any other feature in NBitcoin.

**Comment by NicolasDorier** [comment]
Nuget package can have several libraries? I was unaware of this.

**Comment by knocte** [comment]
Yes, you just need to list several `<file>` elements in the nuspec file. An example: https://github.com/nblockchain/JsonRpcSharp/blob/master/build/JsonRpcSharp.nuspec

**Comment by NicolasDorier** [comment]
Ok. So the idea would be including F# lib inside the NBitcoin nuget package and rely on extensions method?

This seems good.

**Comment by NicolasDorier** [comment]
If we go extension method route, which namespace `NBitcoin.Miniscript` should put the extension methods?

`NBitcoin` seems the best for discovery. `NBitcoin.Extensions` seems best to follow Microsoft best practicies.

**Comment by NicolasDorier** [comment]
@knocte I don't think `<file>` will automatically reference the assembly.

**Comment by NicolasDorier** [comment]
There might have another solution like ILMerge. But afaik this does not work on .NET Core.

**Comment by knocte** [comment]
> I don't think <file> will automatically reference the assembly.

It does it for me. Just try it now: create a C# program and add the nuget package "JsonRpcSharp" to it as a dependency. Many dlls will be linked.

**Comment by NicolasDorier** [comment]
Seems like a good idea then.

**Comment by knocte** [comment]
The only thing that I would recommend to @joemphilips is to migrate from paket to nuget, mixing them is not a good idea IMNSHO.

**Comment by NicolasDorier** [comment]
What about integrating @joemphilips library as a git submodule? (If you agree with renaing it to NBitcoin.MiniScript)
He would just have to do PR here to bump the version.

**Comment by NicolasDorier** [comment]
@knocte I think separate library dll might not work because there is a circular dependency.

**Comment by knocte** [comment]
> I think separate library dll might not work because there is a circular dependency.

I just spotted that, but I don't think this is an unfixable problem. Many other libraries I know have faced these similar growing pains, and therefore I propose two alternatives that I know of:
a) Separate the bits that NBitcoin.MiniScript needs from NBitcoin into a base library. This could be named something like `NBitcoin.Base` or `NBitcoin.Core` or `NBitcoin.Interfaces`. Then NBitcoin could depend on both this library and NBitcoin.MiniScript.
b) Not use extension methods. When a dev needs to consume NBitcoin.MiniScript facilities, he will just need to use the proper namespace of it.

**Comment by NicolasDorier** [comment]
I don't want to break NBitcoin in several libs. This is a breaking change. Maybe some hack with `ILLink` might make it possible to integrate Miniscript directly into NBitcoin while keeping it F#

**Comment by NicolasDorier** [comment]
Actually we are silly, this should be fine as git module.
Miniscript depends on NBitcoin, not the reverse, so there is no circular deps.

**Comment by knocte** [comment]
> Miniscript depends on NBitcoin, not the reverse, so there is no circular deps.

This is what I was thinking from the beginning, but as you said you were mentioning extension methods and wanting NBitcoin.dll to depend on NBitcoin.MiniScript.dll, I thought that you just wanted MiniScript API to be found via intellisense when exploring the methods of current NBitcoin Script type, or something like that.

> this should be fine as git module.

I would avoid submodules if you can avoid them, they are a pain to deal with. Why not just a folder in NBitcoin repo?

**Comment by NicolasDorier** [comment]
Yeah you are right, no good reason.

**Comment by joemphilips** [comment]
So should I just make a PR which creates another NuGet project named `NBitcoin.Miniscript` in this repository without touching main repo as much as possible? I would like to be clarified.

Anyway, I will keep coding. Decompiler seems to be complicated than I first expected.

**Comment by NicolasDorier** [comment]
@joemphilips maybe keep it in C#, put it in this repo, and find a way so that we can add it in the NBitcoin nuget package.

Add extensions methods in the NBitcoin namespace of your project so your methods are discoverable.

**Comment by knocte** [comment]
>@joemphilips maybe keep it in C#

Note that was a typo, Nicolas meant F#

> which creates another NuGet project named NBitcoin.Miniscript

No, we were saying that more than 1 dll can be included in the same project. Just create a new `<file>` entry for your lib in the nuspec file.

**Comment by joemphilips** [comment]
Closing in favor of #675
